sdk/core/System.ClientModel/src/Options/RequestOptions.cs

```diff
@@ -87,10 +98,24 @@ internal void Apply(PipelineMessage message)
         message.PerTryPolicies = _perTryPolicies;
         message.BeforeTransportPolicies = _beforeTransportPolicies;
 
-        // Add the values of any headers set via AddHeader.
-        foreach (var header in _addHeaders)
+        // Apply adds and sets to request headers if applicable.
+        if (_headersUpdates is not null)
         {
-            message.Request.Headers.Add(header.Key, header.Value);
+            foreach (var update in _headersUpdates)
+            {
+                switch (update.Operation)
+                {
+                    case HeaderOperation.Add:
+
+                        message.Request.Headers.Add(update.HeaderName, update.HeaderValue);
+                        break;
+                    case HeaderOperation.Set:
+                        message.Request.Headers.Set(update.HeaderName, update.HeaderValue);
+                        break;
```

Comment:
I'm curious if the distinction between `Add` and `Set` is important here. I assume that `Add` would set if the header existed and `Set` would add if it didn't. Should we simplify to something like `AddorSetHeader` for the API?

Reply:
Add will append a value to any existing value, whereas Set has replace semantics.  It's important to differentiate between them because if people call them in different orders, they can get a different result.  I believe other languages support differentiating between the two, so adding SetHeader to RequestOptions follows precedent across languages.

Does that address your question, or was there a case you were thinking of that I didn't understand?

Reply:
> Does that address your question, or was there a case you were thinking of that I didn't understand?

OK - I didn't consider the append scenario for add. Is it correct then that the only way to have a header with the two values "foo, bar" is to call `AddHeader` twice? If so, then the process to update such a header to "foo, barNew" would be to call `SetHeader("myHeader", "foo")` and `AddHeader("myHeader", "barNew")`?

My thinking was that this seems more complicated than if we supported something like the following instead:

`AddOrSetHeader(string name, string value);`
`AddOrSetHeader(string name, IEnumerable<string> value);`

Reply:
You could also call one of the following:
options.SetHeader("myHeader", "foo,barNew");   // replace any existing headers
options.AddHeader("myHeader", "foo,barNew"); // append the values to any current values for the header

It would be possible to add overloads for the IEnumerable in the future if needed, but I think it's ok to start small here.

Reply:
This is largely consistent with the [PipelineRequestHeaders](https://apiview.dev/Assemblies/Review/1b123e7a51d44ebe945f0212ee039c65#System.ClientModel.Primitives.PipelineRequestHeaders) APIs we have today, so I'd rather avoid adding methods that diverge from that to add new semantics.

Reply:
> You could also call one of the following:
> 
> ```cs
> options.SetHeader("myHeader", "foo,barNew"); // replace existing values for "myHeader"
> options.AddHeader("myHeader", "foo,barNew"); // append the values to any current values for the "myHeader"
> ```

If those two examples are equivalent, then I think I am misunderstanding what is meant by append here. My interpretation is that calling `options.AddHeader("myHeader", "foo,barNew")` if the current value of myHeader was "foo,bar" would result in the new value being "foo,bar,foo,barNew".

Reply:
No, they're not equivalent, you're right about that.  If you call them in that sequence you are correct that the result would be `"foo,barNew,foo,barNew".  My goal of sharing those was to answer your question that you did not have to call `AddHeader` twice.

Reply:
Discussed offline:
- @christothes's specific concern is that he's not sure how common the Add semantics are, and that it might cause confusion to have both
- IIRC, in API discussions @KrzysztofCwalina and @tg-msft felt that having both side-by-side would actually help reduce confusion, by giving users a clue that there are different semantics for Add and Set.  We had also discussed in the Jan 11 arch board meeting that other SDK languages provide both APIs.

@christothes, please feel free to add if this isn't a valid summary!